### PR TITLE
etcdhttp: add health endpoint

### LINKED ISSUE #2040: health check endpoint
An explicit health check endpoint can be added to check whether the cluster is working well.
The endpoint makes it easy to reason about when debugging or monitoring.

from #929 and @jakedt and @robszumski

### PR DESCRIPTION
**Pull request by xiang90**
Fix #2040

### CODE REVIEW

File: etcdserver/etcdhttp/client.go [hunk 1]
```diff
@@ -74,6 +75,7 @@ func NewClientHandler(server *etcdserver.EtcdServer) http.Handler {
 
 	mux := http.NewServeMux()
 	mux.HandleFunc("/", http.NotFound)
+	mux.Handle("/health", healthHandler(server))
```
Comment (yichengq) [inline etcdserver/etcdhttp/client.go]: use `healthPrefix`

File: etcdserver/etcdhttp/client.go [hunk 2]
```diff
@@ -281,6 +283,35 @@ func (h *statsHandler) serveLeader(w http.ResponseWriter, r *http.Request) {
...
+// TODO: change etcdserver to raft interface when we have it.
+//       add test for healthHeadler when we have the interface ready.
+func healthHandler(server *etcdserver.EtcdServer) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		if !allowMethod(w, r.Method, "GET") {
+			return
+		}
+
+		if uint64(server.Leader()) == raft.None {
+			http.Error(w, `"{health": "false"}`, http.StatusServiceUnavailable)
```
Comment (yichengq) [inline etcdserver/etcdhttp/client.go]: `{"health": "false"}`

File: etcdserver/etcdhttp/client.go [hunk 3]
```diff
@@ -281,6 +283,35 @@ func (h *statsHandler) serveLeader(w http.ResponseWriter, r *http.Request) {
...
+			return
+		}
+
+		// wait for raft's progress
+		index := server.Index()
+		for i := 0; i < 3; i++ {
+			time.Sleep(250 * time.Millisecond)
```
Comment (yichengq) [inline etcdserver/etcdhttp/client.go]: could we use a const for sync interval?

Comment (xiang90) [inline etcdserver/etcdhttp/client.go]: Not worth it. We do not care about the sync interval. The general idea is etcd should make a progress within 0.5 second.

Comment (yichengq) [inline etcdserver/etcdhttp/client.go]: k

File: etcdserver/etcdhttp/client.go [hunk 4]
```diff
@@ -281,6 +283,35 @@ func (h *statsHandler) serveLeader(w http.ResponseWriter, r *http.Request) {
...
+			if server.Index() > index {
+				w.WriteHeader(http.StatusOK)
+				w.Write([]byte(`"{health": "true"}`))
```
Comment (yichengq) [inline etcdserver/etcdhttp/client.go]: {"health": "true"}

File: etcdserver/etcdhttp/client.go [hunk 5]
```diff
@@ -281,6 +283,35 @@ func (h *statsHandler) serveLeader(w http.ResponseWriter, r *http.Request) {
...
+				return
+			}
+		}
+
+		http.Error(w, `"{health": "false"}`, http.StatusServiceUnavailable)
```
Comment (yichengq) [inline etcdserver/etcdhttp/client.go]: {"health": "false"}

**Comment by yichengq** [comment]
lgtm

**Pull request merged**